File: Packages/MumbleCore/Sources/MumbleProtocol/OCB2.swift
import Foundation

public protocol BlockCipher {
    func encryptBlock(_ input: [UInt8]) -> [UInt8]
    func decryptBlock(_ input: [UInt8]) -> [UInt8]
}

public final class CryptState {
    public static let blockSize = 16
    private static let headerBytes = 4
    private static let reorderWindow = 30

    private var cipher: BlockCipher?
    private(set) public var encryptIV = CryptState.zeroBlock
    private(set) public var decryptIV = CryptState.zeroBlock
    private var decryptHistory = [UInt8](repeating: 0, count: 256)

    public private(set) var good: UInt32 = 0
    public private(set) var late: UInt32 = 0
    public private(set) var lost: UInt32 = 0
    public private(set) var resync: UInt32 = 0
    public private(set) var lastGood: Date = .distantPast
    public private(set) var lastRequest: Date = .distantPast

    public var isValid: Bool { cipher != nil }

    private let makeCipher: ([UInt8]) -> BlockCipher

    public init(cipherFactory: @escaping ([UInt8]) -> BlockCipher = { AES128(key: $0) }) {
        self.makeCipher = cipherFactory
    }

    private static var zeroBlock: [UInt8] { [UInt8](repeating: 0, count: blockSize) }

    public func setKey(_ key: [UInt8], encryptIV: [UInt8], decryptIV: [UInt8]) -> Bool {
        let blockSize = Self.blockSize
        guard key.count == blockSize, encryptIV.count == blockSize, decryptIV.count == blockSize else { return false }
        self.encryptIV = encryptIV
        self.decryptIV = decryptIV
        cipher = makeCipher(key)
        decryptHistory = [UInt8](repeating: 0, count: 256)
        good = 0
        late = 0
        lost = 0
        resync = 0
        return true
    }

    public func setKey(_ key: Data, clientNonce: Data, serverNonce: Data) -> Bool {
        setKey([UInt8](key), encryptIV: [UInt8](clientNonce), decryptIV: [UInt8](serverNonce))
    }

    public func setDecryptIV(_ iv: [UInt8]) -> Bool {
        guard iv.count == Self.blockSize else { return false }
        decryptIV = iv
        resync += 1
        return true
    }

    public func setEncryptIV(_ iv: [UInt8]) -> Bool {
        guard iv.count == Self.blockSize else { return false }
        encryptIV = iv
        return true
    }

    public func markResyncRequested() {
        lastRequest = Date()
    }

    public func encrypt(_ plain: Data) -> Data? {
        guard cipher != nil else { return nil }
        Self.increment(&encryptIV, from: 0)
        let (ciphertext, tag) = ocbEncrypt([UInt8](plain), nonce: encryptIV)
        var packet = [encryptIV[0], tag[0], tag[1], tag[2]]
        packet.append(contentsOf: ciphertext)
        return Data(packet)
    }

    public func decrypt(_ crypted: Data) -> Data? {
        guard cipher != nil, crypted.count >= Self.headerBytes else { return nil }
        let source = [UInt8](crypted)
        let saved = decryptIV
        guard let adjustment = advanceDecryptIV(to: source[0]) else {
            decryptIV = saved
            return nil
        }
        let (plain, tag, ok) = ocbDecrypt(Array(source[Self.headerBytes...]), nonce: decryptIV)
        let tagMatches = tag[0] == source[1] && tag[1] == source[2] && tag[2] == source[3]
        guard ok, tagMatches else {
            decryptIV = saved
            return nil
        }
        decryptHistory[Int(decryptIV[0])] = decryptIV[1]
        if adjustment.restore { decryptIV = saved }
        good &+= 1
        late = Self.adjusted(late, by: adjustment.late)
        lost = Self.adjusted(lost, by: adjustment.lost)
        lastGood = Date()
        return Data(plain)
    }

    private func advanceDecryptIV(to ivByte: UInt8) -> (restore: Bool, lost: Int, late: Int)? {
        let current = decryptIV[0]
        if current &+ 1 == ivByte {
            if ivByte > current {
                decryptIV[0] = ivByte
            } else if ivByte < current {
                decryptIV[0] = ivByte
                Self.increment(&decryptIV, from: 1)
            } else {
                return nil
            }
            return (restore: false, lost: 0, late: 0)
        }
        var diff = Int(ivByte) - Int(current)
        if diff > 128 {
            diff -= 256
        } else if diff < -128 {
            diff += 256
        }
        let isLate = diff > -Self.reorderWindow && diff < 0
        var restore = false
        var lost = 0
        var late = 0
        if ivByte < current && isLate {
            late = 1
            lost = -1
            decryptIV[0] = ivByte
            restore = true
        } else if ivByte > current && isLate {
            late = 1
            lost = -1
            decryptIV[0] = ivByte
            Self.decrement(&decryptIV, from: 1)
            restore = true
        } else if ivByte > current && diff > 0 {
            lost = Int(ivByte) - Int(current) - 1
            decryptIV[0] = ivByte
        } else if ivByte < current && diff > 0 {
            lost = 256 - Int(current) + Int(ivByte) - 1
            decryptIV[0] = ivByte
            Self.increment(&decryptIV, from: 1)
        } else {
            return nil
        }
        if decryptHistory[Int(decryptIV[0])] == decryptIV[1] { return nil }
        return (restore: restore, lost: lost, late: late)
    }

    private static func adjusted(_ counter: UInt32, by delta: Int) -> UInt32 {
        if delta > 0 { return counter &+ UInt32(delta) }
        if Int(counter) > -delta { return counter - UInt32(-delta) }
        return counter
    }

    private static func increment(_ iv: inout [UInt8], from start: Int) {
        for index in start..<blockSize {
            iv[index] &+= 1
            if iv[index] != 0 { break }
        }
    }

    private static func decrement(_ iv: inout [UInt8], from start: Int) {
        for index in start..<blockSize {
            let before = iv[index]
            iv[index] &-= 1
            if before != 0 { break }
        }
    }

    static func double(_ block: inout [UInt8]) {
        let carry = block[0] >> 7
        for index in 0..<(blockSize - 1) {
            block[index] = (block[index] << 1) | (block[index + 1] >> 7)
        }
        block[blockSize - 1] = (block[blockSize - 1] << 1) ^ (carry &* 0x87)
    }

    static func triple(_ block: inout [UInt8]) {
        var doubled = block
        double(&doubled)
        for index in 0..<blockSize { block[index] ^= doubled[index] }
    }

    @inline(__always)
    private static func xor(_ left: [UInt8], _ right: [UInt8]) -> [UInt8] {
        var out = zeroBlock
        for index in 0..<blockSize { out[index] = left[index] ^ right[index] }
        return out
    }

    private static func lengthBlock(_ byteCount: Int) -> [UInt8] {
        var block = zeroBlock
        let bits = UInt32(byteCount * 8)
        block[12] = UInt8((bits >> 24) & 0xFF)
        block[13] = UInt8((bits >> 16) & 0xFF)
        block[14] = UInt8((bits >> 8) & 0xFF)
        block[15] = UInt8(bits & 0xFF)
        return block
    }

    private static func isZeroExceptLastByte(_ block: [UInt8]) -> Bool {
        var sum: UInt8 = 0
        for index in 0..<(blockSize - 1) { sum |= block[index] }
        return sum == 0
    }

    private static func prefixMatches(_ left: [UInt8], _ right: [UInt8]) -> Bool {
        for index in 0..<(blockSize - 1) where left[index] != right[index] { return false }
        return true
    }

    public func ocbEncrypt(_ plain: [UInt8], nonce: [UInt8]) -> ([UInt8], [UInt8]) {
        guard let cipher else { return ([], Self.zeroBlock) }
        let blockSize = Self.blockSize
        var delta = cipher.encryptBlock(nonce)
        var checksum = Self.zeroBlock
        var out = [UInt8](repeating: 0, count: plain.count)
        var remaining = plain.count
        var offset = 0

        while remaining > blockSize {
            let block = Array(plain[offset..<(offset + blockSize)])
            let flipBit = remaining - blockSize <= blockSize && Self.isZeroExceptLastByte(block)
            Self.double(&delta)
            var masked = Self.xor(delta, block)
            if flipBit { masked[0] ^= 1 }
            let encrypted = Self.xor(delta, cipher.encryptBlock(masked))
            out.replaceSubrange(offset..<(offset + blockSize), with: encrypted)
            checksum = Self.xor(checksum, block)
            if flipBit { checksum[0] ^= 1 }
            remaining -= blockSize
            offset += blockSize
        }

        Self.double(&delta)
        let pad = cipher.encryptBlock(Self.xor(Self.lengthBlock(remaining), delta))
        var tail = Self.zeroBlock
        for index in 0..<remaining { tail[index] = plain[offset + index] }
        for index in remaining..<blockSize { tail[index] = pad[index] }
        checksum = Self.xor(checksum, tail)
        let encryptedTail = Self.xor(pad, tail)
        for index in 0..<remaining { out[offset + index] = encryptedTail[index] }

        Self.triple(&delta)
        let tag = cipher.encryptBlock(Self.xor(delta, checksum))
        return (out, tag)
    }

    public func ocbDecrypt(_ encrypted: [UInt8], nonce: [UInt8]) -> ([UInt8], [UInt8], Bool) {
        guard let cipher else { return ([], Self.zeroBlock, false) }
        let blockSize = Self.blockSize
        var delta = cipher.encryptBlock(nonce)
        var checksum = Self.zeroBlock
        var out = [UInt8](repeating: 0, count: encrypted.count)
        var remaining = encrypted.count
        var offset = 0

        while remaining > blockSize {
            let block = Array(encrypted[offset..<(offset + blockSize)])
            Self.double(&delta)
            let plain = Self.xor(delta, cipher.decryptBlock(Self.xor(delta, block)))
            out.replaceSubrange(offset..<(offset + blockSize), with: plain)
            checksum = Self.xor(checksum, plain)
            remaining -= blockSize
            offset += blockSize
        }

        Self.double(&delta)
        let pad = cipher.encryptBlock(Self.xor(Self.lengthBlock(remaining), delta))
        var tail = Self.zeroBlock
        for index in 0..<remaining { tail[index] = encrypted[offset + index] }
        tail = Self.xor(tail, pad)
        checksum = Self.xor(checksum, tail)
        for index in 0..<remaining { out[offset + index] = tail[index] }
        let success = !Self.prefixMatches(tail, delta)

        Self.triple(&delta)
        let tag = cipher.encryptBlock(Self.xor(delta, checksum))
        return (out, tag, success)
    }
}

File: web/src/ocb2.js
import crypto from 'node:crypto';

const BLOCK = 16;
const TAG_BYTES = 3;
const HEADER = 1 + TAG_BYTES;
const REORDER_WINDOW = 30;

function aes128(key) {
  const encryptor = crypto.createCipheriv('aes-128-ecb', key, null);
  encryptor.setAutoPadding(false);
  const decryptor = crypto.createDecipheriv('aes-128-ecb', key, null);
  decryptor.setAutoPadding(false);
  return {
    encryptBlock: (block) => new Uint8Array(encryptor.update(block)),
    decryptBlock: (block) => new Uint8Array(decryptor.update(block)),
  };
}

function xor(a, b) {
  const out = new Uint8Array(BLOCK);
  for (let i = 0; i < BLOCK; i++) {
    out[i] = a[i] ^ b[i];
  }
  return out;
}

function double(block) {
  const carry = block[0] >> 7;
  for (let i = 0; i < BLOCK - 1; i++) {
    block[i] = ((block[i] << 1) | (block[i + 1] >> 7)) & 0xff;
  }
  block[BLOCK - 1] = ((block[BLOCK - 1] << 1) ^ (carry * 0x87)) & 0xff;
}

function triple(block) {
  const doubled = Uint8Array.from(block);
  double(doubled);
  for (let i = 0; i < BLOCK; i++) {
    block[i] ^= doubled[i];
  }
}

function lengthBlock(byteCount) {
  const block = new Uint8Array(BLOCK);
  const bits = byteCount * 8;
  block[12] = (bits >>> 24) & 0xff;
  block[13] = (bits >>> 16) & 0xff;
  block[14] = (bits >>> 8) & 0xff;
  block[15] = bits & 0xff;
  return block;
}

function incrementFrom(iv, start) {
  for (let i = start; i < BLOCK; i++) {
    iv[i] = (iv[i] + 1) & 0xff;
    if (iv[i]) {
      break;
    }
  }
}

function decrementFrom(iv, start) {
  for (let i = start; i < BLOCK; i++) {
    const before = iv[i];
    iv[i] = (iv[i] - 1) & 0xff;
    if (before) {
      break;
    }
  }
}

export class CryptState {
  constructor(cipherFactory = aes128) {
    this.makeCipher = cipherFactory;
    this.cipher = null;
    this.encryptIV = new Uint8Array(BLOCK);
    this.decryptIV = new Uint8Array(BLOCK);
    this.history = new Uint8Array(256);
    this.good = 0;
    this.late = 0;
    this.lost = 0;
    this.resync = 0;
    this.lastGood = 0;
  }

  get isValid() {
    return !!this.cipher;
  }

  setKey(key, encryptIV, decryptIV) {
    if (key.length !== BLOCK || encryptIV.length !== BLOCK || decryptIV.length !== BLOCK) {
      return false;
    }
    this.cipher = this.makeCipher(Uint8Array.from(key));
    this.encryptIV = Uint8Array.from(encryptIV);
    this.decryptIV = Uint8Array.from(decryptIV);
    this.history.fill(0);
    this.good = 0;
    this.late = 0;
    this.lost = 0;
    this.resync = 0;
    return true;
  }

  setDecryptIV(iv) {
    if (iv.length !== BLOCK) {
      return false;
    }
    this.decryptIV = Uint8Array.from(iv);
    this.resync++;
    return true;
  }

  encrypt(plain) {
    if (!this.cipher) {
      return null;
    }
    incrementFrom(this.encryptIV, 0);
    const { ciphertext, tag } = this.ocbEncrypt(plain, this.encryptIV);
    const out = new Uint8Array(HEADER + ciphertext.length);
    out[0] = this.encryptIV[0];
    out[1] = tag[0];
    out[2] = tag[1];
    out[3] = tag[2];
    out.set(ciphertext, HEADER);
    return out;
  }

  decrypt(crypted) {
    if (!this.cipher || crypted.length < HEADER) {
      return null;
    }
    const ivByte = crypted[0];
    const saved = Uint8Array.from(this.decryptIV);
    const iv = this.decryptIV;
    let restoreAfter = false;
    let lostDelta = 0;
    let lateDelta = 0;

    if (((iv[0] + 1) & 0xff) === ivByte) {
      if (ivByte > iv[0]) {
        iv[0] = ivByte;
      } else if (ivByte < iv[0]) {
        iv[0] = ivByte;
        incrementFrom(iv, 1);
      } else {
        return null;
      }
    } else {
      let diff = ivByte - iv[0];
      if (diff > 128) {
        diff -= 256;
      } else if (diff < -128) {
        diff += 256;
      }
      const isLate = diff > -REORDER_WINDOW && diff < 0;
      if (ivByte < iv[0] && isLate) {
        lateDelta = 1;
        lostDelta = -1;
        iv[0] = ivByte;
        restoreAfter = true;
      } else if (ivByte > iv[0] && isLate) {
        lateDelta = 1;
        lostDelta = -1;
        iv[0] = ivByte;
        restoreAfter = true;
        decrementFrom(iv, 1);
      } else if (ivByte > iv[0] && diff > 0) {
        lostDelta = ivByte - iv[0] - 1;
        iv[0] = ivByte;
      } else if (ivByte < iv[0] && diff > 0) {
        lostDelta = 256 - iv[0] + ivByte - 1;
        iv[0] = ivByte;
        incrementFrom(iv, 1);
      } else {
        return null;
      }
      if (this.history[iv[0]] === iv[1]) {
        this.decryptIV = saved;
        return null;
      }
    }

    const { plain, tag, ok } = this.ocbDecrypt(crypted.subarray(HEADER), iv);
    const tagMatches = tag[0] === crypted[1] && tag[1] === crypted[2] && tag[2] === crypted[3];
    if (!ok || !tagMatches) {
      this.decryptIV = saved;
      return null;
    }
    this.history[iv[0]] = iv[1];
    if (restoreAfter) {
      this.decryptIV = saved;
    }
    this.good++;
    this.late = adjustCounter(this.late, lateDelta);
    this.lost = adjustCounter(this.lost, lostDelta);
    this.lastGood = Date.now();
    return plain;
  }

  ocbEncrypt(plain, nonce) {
    const cipher = this.cipher;
    const delta = cipher.encryptBlock(nonce);
    let checksum = new Uint8Array(BLOCK);
    const out = new Uint8Array(plain.length);
    let remaining = plain.length;
    let offset = 0;
    while (remaining > BLOCK) {
      const block = plain.subarray(offset, offset + BLOCK);
      const flipBit = remaining - BLOCK <= BLOCK && isZeroExceptLastByte(block);
      double(delta);
      let masked = xor(delta, block);
      if (flipBit) {
        masked[0] ^= 1;
      }
      masked = cipher.encryptBlock(masked);
      out.set(xor(delta, masked), offset);
      checksum = xor(checksum, block);
      if (flipBit) {
        checksum[0] ^= 1;
      }
      remaining -= BLOCK;
      offset += BLOCK;
    }
    double(delta);
    const pad = cipher.encryptBlock(xor(lengthBlock(remaining), delta));
    const tail = new Uint8Array(BLOCK);
    for (let i = 0; i < remaining; i++) {
      tail[i] = plain[offset + i];
    }
    for (let i = remaining; i < BLOCK; i++) {
      tail[i] = pad[i];
    }
    checksum = xor(checksum, tail);
    const encryptedTail = xor(pad, tail);
    for (let i = 0; i < remaining; i++) {
      out[offset + i] = encryptedTail[i];
    }
    triple(delta);
    return { ciphertext: out, tag: cipher.encryptBlock(xor(delta, checksum)) };
  }

  ocbDecrypt(encrypted, nonce) {
    const cipher = this.cipher;
    const delta = cipher.encryptBlock(nonce);
    let checksum = new Uint8Array(BLOCK);
    const out = new Uint8Array(encrypted.length);
    let remaining = encrypted.length;
    let offset = 0;
    while (remaining > BLOCK) {
      double(delta);
      const block = encrypted.subarray(offset, offset + BLOCK);
      const plain = xor(delta, cipher.decryptBlock(xor(delta, block)));
      out.set(plain, offset);
      checksum = xor(checksum, plain);
      remaining -= BLOCK;
      offset += BLOCK;
    }
    double(delta);
    const pad = cipher.encryptBlock(xor(lengthBlock(remaining), delta));
    let tail = new Uint8Array(BLOCK);
    for (let i = 0; i < remaining; i++) {
      tail[i] = encrypted[offset + i];
    }
    tail = xor(tail, pad);
    checksum = xor(checksum, tail);
    for (let i = 0; i < remaining; i++) {
      out[offset + i] = tail[i];
    }
    const ok = !prefixMatches(tail, delta);
    triple(delta);
    return { plain: out, tag: cipher.encryptBlock(xor(delta, checksum)), ok };
  }
}

function isZeroExceptLastByte(block) {
  let sum = 0;
  for (let i = 0; i < BLOCK - 1; i++) {
    sum |= block[i];
  }
  return sum === 0;
}

function prefixMatches(a, b) {
  for (let i = 0; i < BLOCK - 1; i++) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}

function adjustCounter(counter, delta) {
  if (delta > 0) {
    return counter + delta;
  }
  if (counter > -delta) {
    return counter + delta;
  }
  return counter;
}
